### PR TITLE
Rearch: Agent State Encapsulation

### PR DESCRIPTION
**Pull request by collijk**
### Background
A large group of maintainers and contributors met do discuss the architectural challenges associated with the existing codebase. Many much-desired features (building new user interfaces, enabling project-specific agents, enabling multi-agent systems) are bottlenecked by the global state in the system.  We discussed the tradeoffs between an incremental system transition and a big breaking version change and decided to go for the breaking version change.  We justified this by saying:

- We can maintain, in essence, the same user experience as now even with a radical restructuring of the codebase
- Our developer audience is struggling to use the existing codebase to build applications and libraries of their own, so this breaking change will largely be welcome.

This PR sets up the new repository structure so we can work collaboratively in a distributed fashion with reasonable expectations of interfaces.  It will serve as the base branch for our system overhaul.  It keeps all the existing codebase in working order but adds:

- an `autogpt.core` subpackage with nested subpackages for all planned core systems.  Each system subpackage has a `base.py` module in which the interface for the system should be defined
- an `autogpt.new_main` module to serve as an eventual entry point for applications using the updated system.
- a `tests.core` subpackage for _unit testing_ of all base abstractions and implementations of core systems.  The `tests.core` subdirectory structure should be an exact replica of the primary package directory with `autogpt/core/SYSTEM_NAME/MODULE_NAME.py` paired with `tests/core/SYSTEM_NAME/test_MODULE_NAME.py`. 

### Documentation
Some docstring guidance is included in subpackage `init` files and some suggested usage of the abstractions is written in `new_main.py` where the agent is constructed.

**Expectation**: All modules, classes, and functions incoming will have, in order of priority:
- explicit typing to enable static analysis.
- docstrings in (XXX format) for all public functions/methods.
- docstrings in (XXX format) for all private functions methods.

We want to move swiftly, so we can sacrifice a bit of documentation now if we need to.  Use this list to prioritize what you spend time on.

### Test Plan
**Expectation**: All PRs for concrete classes will have unit test coverage for all non-trivial logic.  We can discuss exceptions in PRs, because again, we're trying to balance with speed, but you should make every effort to test logic.  It will help us write better (and more testable) systems.  Feel free to reach out to folks (including me) if you don't have experience with testing frameworks.

### CODE REVIEW

**Comment by p-i-** [comment]
This is a mass message from the AutoGPT core team.
Our apologies for the ongoing delay in processing PRs.
This is because we are re-architecting the AutoGPT core!

For more details (and for infor on joining our Discord), please refer to:
  https://github.com/Significant-Gravitas/Auto-GPT/wiki/Architecting

**Comment by anonhostpi** [comment]
"Big breaking version change." I'm happy to get the version change!

**Comment by anonhostpi** [comment]
Brought up in #3856

File: autogpt/core/planning/base.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    from autogpt.core.plugin.base import PluginManager
+    from autogpt.core.workspace.base import Workspace
+
+
+class LLMModel(enum.StrEnum):
+    # This will be defined probably in the LLM interface module
+    GPT3 = "gpt3"
+    GPT4 = "gpt4"
+
+
```
Comment (ntindle) [inline autogpt/core/planning/base.py]: This absolutely needs to be dynamic

**Comment by anonhostpi** [comment]
So I don't think I saw it in the initial PR description, but does this PR plan to make it possible to save agent states as requested in issues like #430?

**Pull request closed without merging**